hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/upgrade/OMPrepareRequest.java

```diff
@@ -0,0 +1,264 @@
+      ratisStateMachineApplied = (lastRatisCommitIndex >=
+          minRatisStateMachineIndex);
+      LOG.debug("{} Current Ratis state machine transaction index {}.",
+          om.getOMNodeId(), lastRatisCommitIndex);
+
+      if (!(omDBFlushed && ratisStateMachineApplied)) {
+        Thread.sleep(flushCheckInterval.toMillis());
+      }
+    }
+
+    // If the timeout waiting for all transactions to reach the state machine
+    // is exceeded, the exception is propagated, resulting in an error response
+    // to the client. They can retry the prepare request.
+    if (!omDBFlushed) {
+      throw new IOException(String.format("After waiting for %d seconds, " +
+              "OM database flushed index %d which is less than the minimum " +
+              "required index %d.",
+          flushTimeout.getSeconds(), lastOMDBFlushIndex, minOMDBFlushIndex));
+    } else if (!ratisStateMachineApplied) {
+      throw new IOException(String.format("After waiting for %d seconds, " +
+              "Ratis state machine applied index %d which is less than" +
+              " the minimum required index %d.",
+          flushTimeout.getSeconds(), lastRatisCommitIndex,
+          minRatisStateMachineIndex));
+    }
+  }
+
+  /**
+   * Take a snapshot of the state machine at the last index, and purge at
+   * least all log with indices less than or equal to the prepare index.
+   * If there is another prepare request or cancel prepare request,
+   * this one will end up purging that request since it was allowed through
+   * the pre-append prepare gate.
+   * This means that an OM cannot support 2 prepare requests in the
+   * transaction pipeline (un-applied) at the same time.
+   */
+  public static void takeSnapshotAndPurgeLogs(long prepareIndex,
+      RaftServer.Division division) throws IOException {
+    StateMachine stateMachine = division.getStateMachine();
+    long snapshotIndex = stateMachine.takeSnapshot();
+
+    if (snapshotIndex < prepareIndex) {
+      throw new IOException(String.format("OM DB snapshot index %d is less " +
+          "than prepare index %d. Some required logs may not have" +
+          "been persisted to the state machine.", snapshotIndex,
+          prepareIndex));
+    }
+
+    CompletableFuture<Long> purgeFuture =
+        division.getRaftLog().onSnapshotInstalled(snapshotIndex);
+
+    try {
+      long actualPurgeIndex = purgeFuture.get();
+
+      if (actualPurgeIndex != snapshotIndex) {
+        LOG.warn("Actual purge index {} does not " +
+              "match specified purge index {}. ", actualPurgeIndex,
+            snapshotIndex);
+      }
+
+      if (actualPurgeIndex < prepareIndex) {
+        throw new IOException(String.format("Actual purge index %d is less " +
+          "than prepare index %d. Some required logs may not have" +
+            " been removed.", actualPurgeIndex, prepareIndex));
+      }
+    } catch (ExecutionException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)